### PR TITLE
Add new kind of disptacher based on a vector of callbacks

### PR DESCRIPTION
**Pull request by SungmanituHota**
Add a new kind of dispatcher

The callbacks are handled using a Dispatch_Item interface.
The callbacks are store as a vector of Dispatch_Item_Interface'Class.

This work is preliminary for a RESTfull API proposition.

### CODE REVIEW

File: demos/stack_disp/Makefile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+############################################################################
+#                              Ada Web Server                              #
+#                                                                          #
+#                     Copyright (C) 2010-2015, AdaCore                     #
```
Comment (TurboGit) [inline demos/stack_disp/Makefile]: 2016 only

File: demos/stack_disp/stack_disp.gpr [hunk 2]
```diff
@@ -0,0 +1,34 @@
+------------------------------------------------------------------------------
+--                              Ada Web Server                              --
+--                                                                          --
+--                     Copyright (C) 2010-2012, AdaCore                     --
```
Comment (TurboGit) [inline demos/stack_disp/stack_disp.gpr]: 2016 only.

File: demos/stack_soap_disp/Makefile [hunk 3]
```diff
@@ -0,0 +1,29 @@
+############################################################################
+#                              Ada Web Server                              #
+#                                                                          #
+#                     Copyright (C) 2010-2015, AdaCore                     #
```
Comment (TurboGit) [inline demos/stack_soap_disp/Makefile]: 2016 only

File: demos/stack_soap_disp/stack_soap_disp.gpr [hunk 4]
```diff
@@ -0,0 +1,34 @@
+------------------------------------------------------------------------------
+--                              Ada Web Server                              --
+--                                                                          --
+--                     Copyright (C) 2010-2012, AdaCore                     --
```
Comment (TurboGit) [inline demos/stack_soap_disp/stack_soap_disp.gpr]: 2016 only.

File: src/core/aws-dispatchers-stacks.ads [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+package AWS.Dispatchers.Stacks is
+
+   type Dispatch_Item_Interface is interface;
+   function Callback (Object : Dispatch_Item_Interface;
+                      Request : AWS.Status.Data)
+                     return AWS.Response.Data is abstract;
+   Not_Handled : exception;
+
+   package Dispatch_Item_Vectors is new Ada.Containers.Indefinite_Vectors
```
Comment (SungmanituHota) [inline src/core/aws-dispatchers-stacks.ads]: Done

File: src/core/aws-dispatchers-stacks.ads [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+--  unit, or you link this unit with other files to produce an executable,  --
+--  this  unit  does not  by itself cause  the resulting executable to be   --
+--  covered by the GNU General Public License. This exception does not      --
+--  however invalidate any other reasons why the executable file  might be  --
+--  covered by the  GNU Public License.                                     --
+------------------------------------------------------------------------------
+
+pragma Ada_2012;
+
+with Ada.Containers.Indefinite_Vectors;
```
Comment (SungmanituHota) [inline src/core/aws-dispatchers-stacks.ads]: Done

File: src/core/aws-dispatchers-stacks.ads [hunk 7]
```diff
@@ -0,0 +1,59 @@
+------------------------------------------------------------------------------
```
Comment (SungmanituHota) [inline src/core/aws-dispatchers-stacks.ads]: Done

File: src/soap/soap-dispatch_item.ads [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+with AWS.Dispatchers.Stacks;
+with AWS.Response;
+with AWS.Status;
+
+with SOAP.Dispatchers;
+with SOAP.WSDL.Schema;
+
+package SOAP.Dispatch_Item is
+
+   type SOAP_Item is
```
Comment (SungmanituHota) [inline src/soap/soap-dispatch_item.ads]: Done

File: demos/stack_disp/Makefile (hunk 1 above)
Comment (SungmanituHota) [inline demos/stack_disp/Makefile]: Done

File: demos/stack_soap_disp/stack_soap_disp.gpr (hunk 4 above)
Comment (SungmanituHota) [inline demos/stack_soap_disp/stack_soap_disp.gpr]: Done

File: src/core/aws-dispatchers-stacks.ads [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+   function Callback (Object : Dispatch_Item_Interface;
+                      Request : AWS.Status.Data)
+                     return AWS.Response.Data is abstract;
+   Not_Handled : exception;
+
+   package Dispatch_Item_Vectors is new Ada.Containers.Indefinite_Vectors
+     (Positive, Dispatch_Item_Interface'Class, "=");
+
+   type Handler is new AWS.Dispatchers.Handler with private;
+   procedure Append_Distpatch_Item (Dispatcher : in out Handler;
```
Comment (SungmanituHota) [inline src/core/aws-dispatchers-stacks.ads]: Done in 92243fad2395b47085e7a7926bbefb05c9b7620b

File: demos/stack_disp/stack_disp.gpr (hunk 2 above)
Comment (SungmanituHota) [inline demos/stack_disp/stack_disp.gpr]: Done in 92243fad2395b47085e7a7926bbefb05c9b7620b

File: src/soap/soap-dispatch_item.ads [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+pragma Ada_2012;
+
+with AWS.Dispatchers.Stacks;
+with AWS.Response;
+with AWS.Status;
+
+with SOAP.Dispatchers;
+with SOAP.WSDL.Schema;
+
+package SOAP.Dispatch_Item is
```
Comment (SungmanituHota) [inline src/soap/soap-dispatch_item.ads]: Done in 0bf29f0ce714bdb652c59769173a0410866768b4

File: src/core/aws-dispatchers-stacks.ads [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+with Ada.Containers.Indefinite_Vectors;
+
+package AWS.Dispatchers.Stacks is
+
+   type Dispatch_Item_Interface is interface;
+   function Callback (Object : Dispatch_Item_Interface;
+                      Request : AWS.Status.Data)
+                     return AWS.Response.Data is abstract;
+   Not_Handled : exception;
```
Comment (SungmanituHota) [inline src/core/aws-dispatchers-stacks.ads]: Done in 46b9f494d01d00cc4db18c01e70d92fff81511eb

File: src/soap/soap-dispatch_item.adb [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+         begin
+            return Object.SOAP_Callback
+              (SOAPAction,
+               SOAP.Message.XML.Load_Payload
+                 (Unbounded_String'(AWS.Status.Payload (Request)),
+                  Schema => Schema (SOAP_Item'Class (Object), SOAPAction)),
+               Request);
+         end;
+      else
+         raise AWS.Dispatchers.Stacks.Not_Handled;
```
Comment (SungmanituHota) [inline src/soap/soap-dispatch_item.adb]: Done in 46b9f494d01d00cc4db18c01e70d92fff81511eb

**Comment by TurboGit** [comment]
Ok, I would squash the 3 batch preview commit into the original one. I think the specs need some comments.

File: src/core/aws-dispatchers-stacks.ads [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+------------------------------------------------------------------------------
+
+pragma Ada_2012;
+
+with Ada.Containers.Indefinite_Vectors;
+
+package AWS.Dispatchers.Stacks is
+
+   type Dispatch_Item_Interface is interface;
+   function Callback (Object : Dispatch_Item_Interface;
```
Comment (TurboGit) [inline src/core/aws-dispatchers-stacks.ads]: This is not really a Callback but a service on the interface. Not sure what a better name could be?
Maybe just Dispatch as in fact it dispatch on the item set?

Comment (SungmanituHota) [inline src/core/aws-dispatchers-stacks.ads]: I have changed the name to Item_Interface in the package
src/extended/aws-services-dispatchers-stack.ads

Is this OK for you ?

**Comment by SungmanituHota** [comment]
I'm using a mercurial repository with hg-git. I don't know if I can do this kind of thing.

**Comment by SungmanituHota** [comment]
May be I can use something like that : 
https://www.mercurial-scm.org/wiki/ConcatenatingChangesets

**Comment by SungmanituHota** [comment]
Maybe I can try on a git repository version of aws.git and then clone it to a fresh mercurail repo to continue my work.

What's the usual command ? A rebase one ?
Sorry for the question, I'm really not a Git user.

--
xavier

----- Mail original -----

**Comment by TurboGit** [comment]
If you are not a Git user that's an advanced use and not that easy :)

The idea is to do an interactive rebase:
$ git rebase --interactive HEAD~n

Where N is the number of commit on your current branch to rebase. You'll be given with an editor and the first column will be the action to do on the commit. The 's' or 'f' fixup command can be used to merge commits together. You can also reorder the commit in the editor.

All this is to give a clean topic branch before merging. All the stuff we are discussing, fixing are really not to be exposed on the final merge.

If you cannot do that I can do it on a topic branch on my AWS GitHub remote. That will be quite straightforward for me. Let me know...

**Comment by SungmanituHota** [comment]
If this is OK for you to do it in a topic branch - please do so. 
If I find a git guru in one of our companion laboratories may be I'll try it .

**Comment by TurboGit** [comment]
The license was never signed. I'll close please feel free to reopen or create a new PR if needed. And sorry for the (very) late reply.

**Pull request closed without merging**